plugins/outputs/zabbix/utils.go

```diff
@@ -0,0 +1,113 @@
+package zabbix
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/datadope-io/go-zabbix/v2"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+)
+
+var (
+	errMetricNoMeasurementName = fmt.Errorf("empty measurement name")
+	errMetricNoTags            = fmt.Errorf("metric without tags")
+	errMetricJustHostTag       = fmt.Errorf("metric with only host tag")
+)
+
+// getHostname returns the hostname from the tags, or the system hostname if not found.
+func getHostname(hostTag string, metric telegraf.Metric) (string, error) {
+	if hostname, ok := metric.GetTag(hostTag); ok {
+		return hostname, nil
+	}
+
+	return os.Hostname()
+}
+
+// processMetric converts a Telegraf metric to a list of Zabbix metrics.
+// Ignore metrics with no hostname.
+func (z Zabbix) processMetric(metric telegraf.Metric) []*zabbix.Metric {
+	zbxMetrics := make([]*zabbix.Metric, 0, len(metric.FieldList()))
+
+	for _, field := range metric.FieldList() {
+		zbxMetric, err := z.buildZabbixMetric(metric, field.Key, field.Value)
+		if err != nil {
+			z.Log.Errorf("Error converting telegraf metric to Zabbix format: %v", err)
+			continue
+		}
+
+		zbxMetrics = append(zbxMetrics, zbxMetric)
+	}
+
+	return zbxMetrics
+}
+
+// buildZabbixMetric builds a Zabbix metric from a Telegraf metric, for one particular value.
+func (z Zabbix) buildZabbixMetric(metric telegraf.Metric, fieldName string, value interface{}) (*zabbix.Metric, error) {
+	hostname, err := getHostname(z.HostTag, metric)
+	if err != nil {
+		return nil, fmt.Errorf("error getting hostname: %w", err)
+	}
+
+	metricValue, err := internal.ToString(value)
+	if err != nil {
+		return nil, fmt.Errorf("error converting value: %w", err)
+	}
+
+	key := z.Prefix + metric.Name() + "." + fieldName
+	if z.SkipMeasurementPrefix {
+		key = z.Prefix + fieldName
+	}
+
+	// Ignore host tag.
+	// We want to add tags to the key in alphabetical order. Eg.:
+	// lld.dns_query.query_time_ms[DOMAIN,RECORD_TYPE,SERVER]
+	// TagList already return the tags in alphabetical order.
+	tagValues := make([]string, 0, len(metric.TagList()))
+
+	for _, tag := range metric.TagList() {
+		if tag.Key == z.HostTag {
+			continue
+		}
+
+		// Get tag values in the same order as the tag keys in the tags slice.
+		tagValues = append(tagValues, tag.Value)
+	}
+
+	if len(tagValues) != 0 {
+		key = fmt.Sprintf("%v[%v]", key, strings.Join(tagValues, ","))
+	}
+
+	return zabbix.NewMetric(hostname, key, metricValue, z.AgentActive, metric.Time().Unix()), nil
+}
```

Comment:
You might want to move those function into `zabbix.go` to have all "class" functions in one place... But not a must...

Reply:
Fixed in ce7403273